### PR TITLE
Dockerized LLVM

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

BPF programs will soon require customized LLVM toolchain

#### Summary of Changes

Sync and build LLVM in docker and with 2 stage docker image make the binaries available to run directly.  This docker is optionally used to build BPF programs.

Further changes are coming to populate this docker with the llvm binaries build and released separately.

Fixes #

### CODE REVIEW

**Comment by mvines** [comment]
looks good so far, I see the `WIP` label so I'll refrain from nits or looking too hard

**Comment by jackcmay** [comment]
@mvines. The WIP is more to sync up with what you have in mind.  This is working but is setup so that we run llvm from within Docker.  With the Travis building and posting binaries in release tags would we still be building in a docker?

**Comment by mvines** [comment]
> With the Travis building and posting binaries in release tags would we still be building in a docker?

Probably not, ideally we can use the native binaries directly.  But the docker setup still has value in the short term, in that we'd likely first get Travis building Linux LLVM binaries, so macOS could continue to use the docker image until we get Travis building macOS LLVM binaries.   For Windows support I feel that we can just create a new issue for it and one day somebody may be motivated enough to pick it up.

**Comment by jackcmay** [comment]
@borromeotlhs 
We are building the programs (smart contracts) with LLVM.  We need custom changes to LLVM to support things like BPF shared objects.  Up until this point, we've been downloading publically available LLVM binaries.  In this PR we are building our own forks of LLVM.  To make this easier we are doing this in docker and will be pushing the resulting docker images to docker hub.

File: programs/bpf/c/sdk/llvm/llvm-docker/bin/clang [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash -ex
```
Comment (mvines) [inline programs/bpf/c/sdk/llvm/llvm-docker/bin/clang]: `#!/usr/bin/env bash -ex` doesn't work in all environments iirc  gotta use this instead:
```bash
#!/usr/bin/env bash
set -ex
```

File: programs/bpf/c/sdk/bpf.mk [hunk 2]
```diff
@@ -102,8 +107,10 @@ help:
 	@echo ''
 	@echo 'User settings'
 	@echo '  - The following setting are overridable on the command line, default values shown:'
-	@echo '    - Show commands while building:'
-	@echo '      V=1'
+	@echo '    - Show commands while building: V=1'
+	@echo '      V=$(V)'
+	@echo '    - Use LLVM from docker: DOCKER=1'
```
Comment (mvines) [inline programs/bpf/c/sdk/bpf.mk]: probably should document somewhere around here that the user needs to `docker pull solanalabs/llvm`

File: sdk/docker-llvm/README.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
+## Solana Customized LLVM
+
+This Docker contains LLVM binaries that incorporate customizations and fixes required
+by Solana but not yet upstreamed into the LLVM mainline.
+
+https://hub.docker.com/r/solanalabs/llvm/
+
+### Usage:
+
+This Docker is optionally used by the SDK build system.
```
Comment (mvines) [inline sdk/docker-llvm/README.md]: Can we clarify this as the "SDK BPF build system" (until we use it elsewhere)

File: programs/bpf/c/sdk/bpf.mk (hunk 2 above)
Comment (mvines) [reply programs/bpf/c/sdk/bpf.mk]: `ci/docker-run.sh` does this, but there's also cases when you don't want to pull (you're on an airplane with no uplink) so if we autopull then we'd also want to add a way to disable it.  My vote for now is to just document that the user must pull

**Pull request merged**